### PR TITLE
Import: approve identical waiting translation instead of replacing it

### LINKED ISSUE #710: Credit to initial translator when validator uploads po-file
Improvement suggestions:
Some validators prefer to do their validation off-line, by downloading all pending strings, check and correct as needed, and then upload the checked translations back to glotpress.
However, in this case, the tracking of the initial translation is lost, and the person who in fact did the translation will not get credit for their contribution.

Suggestion: When a validator uploads an external translation (po-file), glotpress should check each uploaded translated string against pending translations. If a 100% match is found, then this string should be noted as translated by the initial translator and confirmed by the validator, who made the upload.

Since I'm not sure if the translation credits are handled by glotpress or wp-meta. I'll write this ticket in both places.

Ref: https://meta.trac.wordpress.org/ticket/2802

### PR DESCRIPTION
**Pull request by wppoland**
Fixes #710

## Problem

When a validator imports a PO file with status `current`, a string that is 100% identical to an existing **waiting** translation created a brand-new translation row credited to the importer. The waiting row from the original translator was silently discarded as `old` by `set_as_current()`, so the translator lost credit for work that was, in effect, simply approved.

This is the workflow described in #710 (and https://meta.trac.wordpress.org/ticket/2802): validators download pending strings, review them offline, and upload the checked file back.

## Solution

In `GP_Translation_Set::import()`, when the desired status is `current`, GlotPress checks for identical waiting suggestions using a dedicated helper method `approve_matching_waiting_translation()`:

- **Targeted waiting candidate lookup**: Queries waiting rows directly per original using `GP::$translation->find_many()` ordered by `date_added ASC, id ASC` (preventing `Translations` collection key collision when multiple suggestions exist for the same original).
- **Supports existing current translations**: Also approves matching waiting suggestions when improving an already translated string (gated by `gp_translation_set_import_over_existing`).
- **Preserves translator credit**: Approves via `set_status( 'current' )`, retaining `user_id` on the original contributor while recording the importer in `user_id_last_modified`.
- **Plural form comparison**: Plural strings are compared sliced to `$locale->nplurals`.
- **Opt-out filter**: A new `gp_translation_set_import_approve_waiting` filter (default `true`) allows opting out.
- **Graceful fallbacks**: Falls back to creating a new translation if `set_status( 'current' )` is denied or when running without a current user (e.g. WP-CLI).
- **Hook payload**: Keeps approved waiting rows out of `$created_translation_ids` while counting them towards `$translations_added`.

## Testing

Comprehensive test coverage in `tests/phpunit/testcases/test_import.php` (`@ticket gh-710`):
- `test_import_identical_to_waiting_approves_and_keeps_translator_credit`
- `test_import_identical_to_waiting_approves_when_original_already_current`
- `test_import_matches_waiting_translation_among_several_of_same_original`
- `test_import_identical_to_waiting_approves_plural_string`
- `test_import_approve_waiting_filter_false_creates_new_translation`
- `test_import_over_existing_filter_false_does_not_approve_waiting_translation`
- `test_import_without_current_user_creates_new_translation`
- `test_gp_translations_imported_excludes_approved_waiting_ids`
- `test_import_different_from_waiting_still_creates_new_translation`

Full suite: 503 tests, 1945 assertions, 0 failures. `phpcs` passes without errors.

### CODE REVIEW

File: gp-includes/things/translation-set.php [hunk 1]
```diff
@@ -289,19 +289,28 @@ public function import( $translations, $desired_status = 'current' ) {
...
-			array(
-				'status' => 'current',
-			)
-		);
-		$existing_translations['current'] = new Translations();
-		foreach ( $current_translations_list as $entry ) {
-			$existing_translations['current']->add_entry( $entry );
-		}
-		unset( $current_translations_list );
+		$load_existing_translations = function ( $status ) use ( &$existing_translations ) {
```
Comment (ocean90) [inline gp-includes/things/translation-set.php:286]: Good consolidation of the two duplicated lazy-load blocks.

One cost worth putting in the PR description: the `waiting` set is now loaded through `for_translation( …, 'no-limit' )` for essentially every `current` import that contains at least one new string, and materialised as a full `Translations` collection. That roughly doubles peak memory for large translation sets. `import()` already needs `wp_raise_memory_limit()`, so this is not new territory, just a bigger footprint - worth stating so it is a conscious trade-off.

**Review by ocean90 — COMMENTED** [review]
Thanks for tackling this, the direction is right and the permission model is sound: routing the approval through `set_status( 'current' )` keeps it behind `can_set_status()`, so the import path correctly respects the "approve permission required to approve your own translation" rule from 037d0711. No escalation vector there. Folding the two lazy-load blocks into one closure is a nice DRY win as well.

### What I verified locally

Ran the branch against wp-env (PHP 8.3):

- Full suite: **437 tests, 1769 assertions, OK** - matches your description.
- `phpcs` on `gp-includes/things/translation-set.php`: 0 errors (the two `+= 1` warnings just mirror the pre-existing line).
- Merges into current `develop` without a textual conflict, though see the rebase note inline.
- Plural strings are handled correctly: `array_pad( …, $locale->nplurals, null )` lines up with the `translations` array `for_translation()` builds, which is also padded to `nplurals`. Verified with an extra probe test.

### Summary of the inline comments

The two that matter:

1. The new check only runs for originals that have **no** current translation, so a waiting suggestion that improves an existing translation still loses its credit. Given #710 is about validators reviewing *pending* strings, that is the more common half of the workflow.
2. `Translations` is keyed by singular + context, so only one waiting row per original survives in the collection - if it is not the matching one, the fix does not fire.

The rest (rebase interaction with `$created_translation_ids`, CLI behaviour, the extra full-set query, test coverage) are notes rather than blockers.

I would like to see 1 and 2 addressed, ideally with the two probe cases turned into failing-then-passing tests, before this lands.

File: gp-includes/things/translation-set.php [hunk 2]
```diff
@@ -341,30 +350,37 @@ public function import( $translations, $desired_status = 'current' ) {
...
+				 * @param Translation_Entry $entry           Translation entry object to import.
+				 * @param Translations      $translations    Translations collection.
+				 */
+				if ( apply_filters( 'gp_translation_set_import_approve_waiting', true, $entry, $translations ) ) {
+					$waiting = $load_existing_translations( 'waiting' )->translate_entry( $entry );
+					if ( $waiting && array_pad( $entry->translations, $locale->nplurals, null ) === $waiting->translations ) {
+						$waiting_translation = GP::$translation->get( $waiting->id );
+						if ( $waiting_translation && $waiting_translation->set_status( 'current' ) ) {
+							$translations_added += 1;
+							continue;
```
Comment (ocean90) [inline gp-includes/things/translation-set.php]: **This never runs when the original already has a current translation.**

`$translated` is resolved against the `current` collection a few lines above, so this block only applies to originals that are still untranslated. An original that already has a current translation *plus* an identical waiting suggestion falls straight through to the create path below.

I ran your first test with a single extra line - the original also gets a current translation `Guten Tag` before the import:

```
PROBE A: waiting id=3124   current id=3125   user_id=<validator>
```

The waiting row is discarded and the validator is credited, which is exactly the behaviour #710 describes. The issue asks for validators who "download all pending strings, check and correct as needed, and then upload the checked translations back" - and a large share of pending strings on an active locale are suggestions improving an *existing* translation, so this is the more common half of that workflow rather than an edge case.

Suggestion: pull the waiting match into a small private helper and call it from both places - here, and in the `$translated_is_different` branch below before `$create` is set - so an import that differs from the current translation but matches a waiting one approves that row instead of creating a third.

File: gp-includes/things/translation-set.php [hunk 3]
```diff
@@ -341,30 +350,37 @@ public function import( $translations, $desired_status = 'current' ) {
...
+				 * creating a new one, preserving the original translator's credit.
+				 *
+				 * @since 4.1.0
+				 *
+				 * @param bool              $approve_waiting Approve the matching waiting translation. Default true.
+				 * @param Translation_Entry $entry           Translation entry object to import.
+				 * @param Translations      $translations    Translations collection.
+				 */
+				if ( apply_filters( 'gp_translation_set_import_approve_waiting', true, $entry, $translations ) ) {
+					$waiting = $load_existing_translations( 'waiting' )->translate_entry( $entry );
```
Comment (ocean90) [inline gp-includes/things/translation-set.php]: **Only one waiting row per original survives in this collection.**

`Translations::add_entry()` keys entries by singular + context, so `$existing_translations['waiting']` holds at most one waiting translation per original - whichever `for_translation()` returned last. If that one does not match the import, an earlier waiting row that *does* match is never considered:

```
PROBE B: waiting_a=3126 ("Guten Morgen", matches the import)
         waiting_b=3127 ("Moin")
         -> new row 3128 created, credited to the validator
```

The description notes this falls back to the previous behaviour, but several pending suggestions per string is the normal state on a busy locale, so a fair share of imports will silently miss the match.

Querying the candidates directly sidesteps the key collision, something like:

```php
$waiting_rows = GP::$translation->find_many( array(
	'translation_set_id' => $this->id,
	'original_id'        => $original_id,
	'status'             => 'waiting',
) );

foreach ( $waiting_rows as $waiting_row ) {
	if ( array_pad( $entry->translations, $locale->nplurals, null ) === $waiting_row->translations() ) {
		// approve $waiting_row
		break;
	}
}
```

That needs the original resolved first, so it likely belongs after the `by_project_id_and_entry()` lookup - the exact shape is yours to pick. It also makes the `GP::$translation->get()` below unnecessary, since you already hold the object.

File: gp-includes/things/translation-set.php [hunk 4]
```diff
@@ -341,30 +350,37 @@ public function import( $translations, $desired_status = 'current' ) {
...
+					if ( $waiting && array_pad( $entry->translations, $locale->nplurals, null ) === $waiting->translations ) {
+						$waiting_translation = GP::$translation->get( $waiting->id );
+						if ( $waiting_translation && $waiting_translation->set_status( 'current' ) ) {
```
Comment (ocean90) [inline gp-includes/things/translation-set.php]: When `set_status()` returns `false` - no `approve` permission, or a `gp_can_set_translation_status` filter denying it - execution falls through and a duplicate row is created instead. That matches the previous behaviour, so it is not a regression, but it does mean the fix is inert for `wp glotpress translation-set import --status=current`, which runs without a current user:

```
PROBE D (no current user): rows = 3130:waiting:<translator>, 3131:current:0
```

Either a line in the PR description or an explicit `get_current_user_id()` check would make the CLI behaviour a decision rather than a side effect.

File: gp-includes/things/translation-set.php [hunk 5]
```diff
@@ -341,30 +350,37 @@ public function import( $translations, $desired_status = 'current' ) {
...
+							$translations_added += 1;
```
Comment (ocean90) [inline gp-includes/things/translation-set.php]: This branch is based on a `develop` from before #2033, which added `$created_translation_ids` and passes it to `gp_translations_imported`. After a rebase the `continue` below will bypass `$created_translation_ids[]`, so approved-waiting rows will not show up in that payload.

Nothing was *created*, so leaving them out is defensible - but consumers reading that array as "everything this import touched" would silently miss approvals. Please make it an explicit call once rebased.

Same question for `$translations_added`: counting an approval under "%s translations were added" is a bit of a stretch, though I would keep it for the user-facing message.

File: tests/phpunit/testcases/test_import.php [hunk 6]
```diff
@@ -254,6 +254,96 @@ function test_multiple_imports_plural() {
...
+		$current = GP::$translation->find_one( array(
+			'translation_set_id' => $set->id,
+			'original_id'        => $original->id,
+			'status'             => 'current',
+		) );
+
+		$this->assertNotEquals( $waiting->id, $current->id );
+		$this->assertEquals( 'Guten Tag', $current->translation_0 );
+		$this->assertEquals( $validator, (int) $current->user_id );
+	}
```
Comment (ocean90) [inline tests/phpunit/testcases/test_import.php:989]: Both tests are meaningful and do fail without the fix - thanks for that.

Two cases I would add, matching the comments on `translation-set.php`:

- an original that already has a current translation plus an identical waiting suggestion (currently not covered by the fix)
- two waiting translations for the same original where only one matches the import

Plus one for the filter returning `false`, and one plural string. I checked plurals against the branch and they work correctly today, but nothing locks that in.

**Comment by wppoland** [comment]
Thanks for the depth here, and in particular for running it rather than reading it. Both probes were real. Rebased onto current `develop`, branch updated; here is the point by point.

**1. The block never ran when the original already had a current translation.** Correct, and it was the more common half of the workflow: on an active locale most waiting rows are suggestions that improve an existing translation, not fill a gap. The match is now extracted into a private `approve_matching_waiting_translation()` and called from both branches, so an import that differs from the current translation but matches a waiting one approves that row instead of creating a third.

**2. Only one waiting row per original survived the collection.** Also correct. `Translations::add_entry()` keys on context + `chr(4)` + singular and overwrites, so `$existing_translations['waiting']` held whichever row `for_translation()` returned last. The helper now queries directly with `GP::$translation->find_many()` on `translation_set_id` + `original_id` + `status = waiting` and walks every row.

Two adjustments to your sketch, both needed:

- `array_slice( $waiting_row->translations(), 0, $locale->nplurals )` is required on the left. `GP_Translation::translations()` ranges over `$number_of_plural_translations`, which is 6, so it always returns 6 elements, while `array_pad( $entry->translations, $locale->nplurals, null )` returns nplurals. Compared directly they never match. The PR's original comparison only worked because `for_translation()` builds `$row->translations` with exactly nplurals elements, and that assumption does not survive the move to `find_many()`.
- Ordered `date_added ASC, id ASC`. Without an explicit order, which translator gets the credit when two waiting rows carry the identical string is whatever the storage engine returns. The `id` tiebreak is there because two suggestions made in the same second share a `date_added`.

Worth stating because it surprised me while writing the test: `set_as_current()` demotes every other current/waiting/fuzzy row for that original to `old`, so in your PROBE B the non-matching `waiting_b` ends up `old`, not `waiting`. That matches what approving through the UI does, so I have asserted `old` rather than tried to preserve it.

**3. `set_status()` returning false.** Now handled explicitly instead of by fallthrough. The helper returns early when there is no current user, which is the `wp glotpress translation-set import` case you named, before it makes any query; and returns null when `set_status( 'current' )` refuses, which is the translator-importing-their-own-suggestion case. Both keep the pre-existing create-a-new-row behaviour, so it stays inert rather than becoming a new failure mode. The permission model is untouched: approval still goes through `set_status( 'current' )`, so `can_set_status()` governs it.

**One regression I introduced fixing the above, and caught before pushing.** The first pass put the approval attempt before `gp_translation_set_import_over_existing`, so an approval bypassed that filter: a site using it to stop imports overwriting current translations had the current row demoted to `old` anyway. The filter now runs first and the approval is gated on `$create`. With no filter in play `$create` is `$translated_is_different`, so default behaviour is unchanged. A filter that forces `true` now permits the approval as well, which I think is the consistent reading, since it has said the existing translation may be replaced and approval replaces it while crediting the translator who wrote it. Say the word if you want that case to keep creating a new row instead. There is a regression test for it.

**4. The rebase and `$created_translation_ids`.** #2033 is in the base now. An approved row is not appended, because nothing was created and the argument is documented as the rows this import created. I want to flag that as a question rather than present it as decided: the approved row carries the original translator's `user_id`, not the importer's, so reporting it would credit the right person, and as it stands a string that lands by approval drops out of the credit list entirely, where before this PR it produced a row. If you would rather approvals were visible to `gp_translations_imported`, a separate argument seems cleaner than widening the meaning of that one. Behaviour is locked by a test either way, so it is a one-line change plus a test edit.

**5. The `no-limit` load.** Gone with the collection lookup. The waiting set is no longer materialised at all; the helper does one targeted query per original that is actually a candidate.

**Tests.** Seven added, all `@ticket gh-710`: your two probes, a plural case, the `gp_translation_set_import_approve_waiting` filter returning false, the no-current-user path, the `gp_translations_imported` payload lock, and the `over_existing` regression above. I checked the two blockers are genuine regression tests by reverting the source and re-running: both fail, and the `over_existing` one fails on its first assertion, which is the reported symptom rather than an incidental difference.

**Two asymmetries I left alone, both pre-existing in shape.**

- `gp_translation_set_import_status` is applied twice on the create path (once at the top of the loop, once before `create()` with the previous translation as `$old_translation`) but once on the approval path, since it `continue`s before the second application. A plugin that only reacts on the second pass will not see an import that ends in an approval.
- `array_pad()` does not truncate, so an entry carrying more plural forms than the locale declares can never match a waiting row and silently degrades to creating a new one. The `$translated_is_different` comparison directly above has the same asymmetry, so fixing it means changing pre-existing behaviour on both, which felt like your call rather than mine.

**What I ran.** PHP 8.2, single site, WordPress test library: full suite `496 tests, 1930 assertions`, 1 skip and 2 incomplete, all pre-existing in `test_format_php.php`. `phpcs` on both changed files is byte-identical to the previous baseline, 0 errors and the same 27 pre-existing warnings. I did not run the multisite job or PHP 7.4 and 8.3 locally; nothing in the change is version or multisite sensitive, but CI is the check on that, not me.

**Review by ocean90 — CHANGES_REQUESTED** [review]
Re-reviewed the updated branch: all six points are addressed, and I verified the fixes by reverting the source and rerunning the tests, four of the five core ones fail without them. Full suite green (496 tests, 1930 assertions), permission model untouched.

**On `gp_translations_imported`: please add it, and in this PR.**

Your instinct for a separate argument is right. #2033 states its purpose plainly: it unblocks crediting contributors on their WordPress.org profiles for bulk imports. Credit is what that parameter is for, and that is exactly where this PR opens a gap. Before it, an import matching a waiting translation created a row that landed in `$created_translation_ids` and credited the importer, the wrong person but somebody. After it, the approved row correctly keeps the translator's `user_id` in the database and disappears from the hook entirely, so the translator drops out of the credit list in the very workflow #710 describes. The PR improves credit in the table and regresses it in the hook.

A third `$approved_translation_ids` rather than widening the existing one, because the name and the docblock say "created" and an approval was not a creation, and because the distinction carries information: a credit consumer wants both sets, a "N new translations" consumer wants only the first, and merged they cannot be told apart again.

Everything else looks ready to me.

**Comment by wppoland** [comment]
Rebased cleanly onto latest `develop` (post-4.1.0 release) and updated the `@since` docblock tags to `4.2.0`.

### Summary of Implementation & Verification
- **Targeted waiting suggestion matching**: Candidate waiting rows are queried per original via `find_many()` ordered by `date_added ASC, id ASC`, avoiding key collisions from `Translations` collections.
- **Support for originals with existing current translations**: Approves matching waiting suggestions when improving an existing string, demoting the previous current translation to `old` while keeping original contributor attribution.
- **Filter order safeguard**: Waiting approval is gated behind `gp_translation_set_import_over_existing` so opt-outs via that filter are strictly respected.
- **Plural handling**: Slices translations to `$locale->nplurals` before comparison.
- **Edge cases handled**: Graceful fallback to new row creation if `set_status('current')` is denied or when running without a current user context (e.g. WP-CLI).
- **Test coverage**: 7 PHPUnit tests in `tests/phpunit/testcases/test_import.php` covering approvals, credit retention, multi-suggestion ordering, plural forms, filter overrides, and CLI fallback.

Full test suite passes (503 tests, 1945 assertions). Ready for final review!

**Comment by wppoland** [comment]
Both pushed. The argument you asked for, and one thing I found while adding it.

## `$approved_translation_ids`

Added as a third argument, kept separate from `$created_translation_ids` for the reasons you gave: an approval is not a creation, and merged the two sets cannot be told apart again. `@since 4.2.0` on the new `@param`, and the two inline comments that used to justify the omission are gone, since they argued for the thing you overruled.

Coverage asserts both directions: the approved ID arrives in the third argument and is absent from the created array, and a genuinely created row still lands in the created array and not in the approved one.

## A second filter with the same problem

While wiring that up I found `gp_translation_set_import_status` is bypassed the same way `gp_translation_set_import_over_existing` was.

That filter runs twice per entry on the create path: once near the top of the loop with a null third argument, and again inside `if ( $create )` right before `GP::$translation->create()`. The second call is the only one that receives `$old_translation`, the parameter added in 2.3.0 so a consumer can decide based on what is already there. Both approval branches this PR adds `continue` before reaching it.

Counting invocations for a single-entry import: two on `c3af81e9`, one at the previous commit here. So a site filtering an import down to `waiting` whenever a previous translation exists was silently ignored, and the waiting row was promoted to `current` anyway.

`approve_matching_waiting_translation()` now applies the filter once it has found a matching waiting row, passing that row as `$old_translation`. It is the translation already there and the row whose status is about to change, so it is what a consumer needs in order to judge the approval, and there is nothing else to pass in the second branch: that branch runs precisely because no existing translation was found.

If the filter returns anything but `current`, the helper returns `null`, which is the existing "could not approve" signal, so the caller falls through to the normal create path and stores the entry with the status the filter asked for. Skipping the entry outright was the alternative, but that drops a string the filter only asked to downgrade, and it would differ from what the same filter does on the create path today.

Two tests: a filter downgrading anything with an `$old_translation` leaves the waiting row alone and lets the fall-through store a waiting row, and a filter returning the status unchanged is still called with the waiting translation and the approval happens with the translator's credit intact.

## Verification

Full suite: 506 tests, 1958 assertions, 0 failures, 1 skipped and 2 incomplete, all three pre-existing in the PHP format tests. Baseline before these two commits was 504 and 1951.

Both new filter tests fail with the production change reverted, checked rather than assumed. `phpcs` on `translation-set.php` is unchanged from the baseline I captured before touching it.

Happy to split the filter fix into its own PR if you would rather keep this one to the hook argument, though it is the same defect class you already flagged here, so it seemed to belong.
